### PR TITLE
Fix NPU Qwen3.5 grpo bugs

### PR DESCRIPTION
**Pull request by addsubmuldiv**
# PR type

- [x] Bug Fix
- [ ] New Feature
- [ ] Document Updates
- [x] More Models or Datasets Support

# PR information

This PR enables **Qwen3.5 / Qwen3.5-MoE** GRPO on **NPU**, covering both **colocate** and **server** vLLM rollout paths, and documents the known limitation of **MoE fused expert LoRA** under **PEFT + Transformers 5**.

The changes are grouped into four areas:

```text
1. Qwen3.5 colocate / basic compatibility
2. vLLM-Ascend MoE weight-sync layout compatibility
3. vLLM server-mode weight synchronization compatibility
4. Documentation for PEFT MoE fused expert LoRA limitations
```

This PR only contains framework code changes and does not include local smoke scripts.

## 1. Qwen3.5 RoPE config compatibility with vLLM 0.18 + Transformers 5

Files involved:

- `swift/infer_engine/vllm_engine.py`

### Problem

Qwen3.5 depends on the config / RoPE parameter handling logic in Transformers 5. In the Qwen3.5 config path of vLLM 0.18.x, `ignore_keys_at_rope_validation` may still be passed in as a `list`; however, the Transformers 5 side treats it as a `set` and performs set operations on it, which leads to a type error.

### Fix

When creating the vLLM engine, this PR temporarily patches `PretrainedConfig.convert_rope_params_to_dict()` so that it only converts `ignore_keys_at_rope_validation` from `list` to `set` when needed.

### Upstream status

- This is a compatibility issue specific to the combination of **vLLM 0.18.x** and **Transformers 5**.
- Later vLLM releases have adjusted the related implementation in the Qwen3.5 config path.
- References:
  - https://github.com/vllm-project/vllm/issues/36411
  - https://github.com/vllm-project/vllm/pull/37338

## 2. MoE weight synchronization layout compatibility for vLLM-Ascend

Files involved:

- `swift/model/npu_patch/vllm_ascend.py`
- `swift/model/npu_patch/vllm_ascend_moe.py`
- `swift/rlhf_trainers/rollout_mixin.py`
- `swift/rlhf_trainers/utils.py`

### Problem

During GRPO training, training-side weights need to be synchronized to the vLLM rollout side. For Qwen MoE on vLLM-Ascend, MoE expert weights have backend-specific layouts:

- training-side weights may come from HF / FSDP2 / Megatron layouts;
- vLLM-Ascend may convert MoE weights into a runtime layout suitable for NPU grouped matmul;
- FSDP2 can expose Qwen MoE expert weights as fused `gate_up_proj` / `down_proj` tensors;
- directly writing these tensors into vLLM-Ascend `w13_weight` / `w2_weight` can leave the runtime weight layout in the wrong direction.

One observed failure was the first rollout hitting `npu_grouped_matmul` with a hidden-size mismatch, where the activation K dimension was `hidden_size` but the synced `w13_weight` dimension still matched the TP-sharded expert intermediate dimension.

### Fix

- Add an NPU-only patch for the vLLM-Ascend MoE weight loader.
- Treat both `qwen3_moe` and `qwen3_5_moe` as the Qwen MoE sync family.
- Expand FSDP2 fused Qwen MoE expert names into vLLM checkpoint-style `gate_proj` / `up_proj` / `down_proj` names before calling the vLLM loader.
- For FSDP2 colocate runtime sync, write the processed runtime layout directly:
  - `w13_weight`: `[local_experts, hidden, 2 * intermediate_per_tp]`
  - `w2_weight`: `[local_experts, intermediate_per_tp, hidden]`
- Skip `process_weights_after_loading()` only for this FSDP2 Qwen MoE colocate runtime-sync path, because calling post-load again would transpose the just-synced runtime weights back to the wrong direction.
- Keep server full-weight reload on the preprocessed-layout path and still run post-load processing after reload.

### Upstream status

- vLLM-Ascend has a similar RL / MoE transpose-related fix direction:
  - https://github.com/vllm-project/vllm-ascend/pull/7498
- vLLM also has an issue where MoE weights need to be reprocessed after reload:
  - https://github.com/vllm-project/vllm/issues/42821
- Historically, vLLM has also had loader-related issues around `w13_weight` / `w2_weight`:
  - https://github.com/vllm-project/vllm/issues/16842

## 3. Weight synchronization compatibility for vLLM server mode

Files involved:

- `swift/pipelines/infer/rollout.py`
- `swift/rlhf_trainers/vllm_client.py`

### Problem

In server mode, the training process synchronizes weights to an independent vLLM rollout server through a client. For the **Qwen3.5-MoE + vLLM-Ascend** path, weight synchronization needs to ensure that:

- HCCL broadcast uses the correct device context / stream on NPU;
- full weight reload triggers the required post-load processing in vLLM / vLLM-Ascend;
- LoRA adapter sync and full weight sync do not share the same post-processing logic.

### Fix

- Consolidate tensor broadcast logic in both the server client and server worker into `_broadcast_tensor()`.
- Enter `communicator.device` device context and use the current NPU stream only under NPU runtime.
- Keep the original generic broadcast behavior unchanged for non-NPU branches to avoid affecting GPU.
- Add a `/process_weights_after_loading/` call on the server worker side to rebuild the vLLM-Ascend MoE layout after full reload.
- Skip the full-reload post-load logic for LoRA adapter sync.

## 4. Known limitation: PEFT 0.19 + Transformers 5 fused MoE LoRA

This PR does **not** attempt to fix the PEFT fused expert LoRA path in SWIFT. Instead, it documents the limitation in the NPU support documentation.

### Background

Transformers 5 changes the structure of some MoE experts. Previously, each expert typically exposed independent `nn.Linear` modules such as `gate_proj`, `up_proj`, and `down_proj`; in the new structure, these expert weights may instead be fused into a single `nn.Parameter`, for example:

```text
mlp.experts.gate_up_proj
mlp.experts.down_proj
```

Regular LoRA uses PEFT `target_modules` to match `nn.Module`s, while fused MoE experts require PEFT `target_parameters` to match parameter names directly.

### Current assessment

- `target_parameters` is an upstream mechanism provided by PEFT, but there are still unresolved edge cases in combinations involving `lora_dropout`, ZeRO-3 / FSDP, and multi-adapter setups.
- This is not an NPU-specific issue, nor is it a vLLM / HCCL issue.
- To avoid maintaining an unstable upstream capability branch inside SWIFT, this PR reverts the temporary workaround for this issue in `swift/tuners/peft.py`.
- This PR adds a known-limitation note to:
  - `docs/source/BestPractices/NPU-support.md`
  - `docs/source_en/BestPractices/NPU-support.md`

### Recommendation

- By default, Qwen3.5 GRPO scripts should not explicitly pass `model_type`, so as to avoid triggering the PEFT Transformers 5 MoE conversion path.
- If the model configuration itself already triggers that path, prefer full-parameter training or disable the affected LoRA combinations.
- If users really need to train fused expert parameters, they should wait until the upstream PEFT support becomes stable, or use it only when `lora_dropout=0` and the target backend has been separately validated.

### Related upstream context

- PEFT 0.17.0 introduced `target_parameters` to support expert weights represented as `nn.Parameter`, such as in MoE:
  https://github.com/huggingface/peft/releases/tag/v0.17.0
- PEFT 0.17.1 / #2710 fixed a number of `target_parameters` issues, but still retained limitations such as multi-adapter support:
  https://github.com/huggingface/peft/pull/2710
- TRL issue #5222 also records issues around Transformers 5 MoE fused parameters combined with PEFT `target_parameters` / adapters:
  https://github.com/huggingface/trl/issues/5222
- There are still related edge cases involving `ParamWrapper` under ZeRO-3 / FSDP:
  https://github.com/huggingface/peft/issues/2916
  https://github.com/huggingface/peft/issues/3080

# Experiment results

Verified:

1. Python static checks passed:

```bash
python -m py_compile \
  swift/infer_engine/vllm_engine.py \
  swift/model/npu_patch/vllm_ascend.py \
  swift/model/npu_patch/vllm_ascend_moe.py \
  swift/pipelines/infer/rollout.py \
  swift/rlhf_trainers/rollout_mixin.py \
  swift/rlhf_trainers/utils.py \
  swift/rlhf_trainers/vllm_client.py \
  swift/tuners/peft.py
```

2. Documentation and diff format checks passed:

```bash
git diff --check -- \
  docs/source/BestPractices/NPU-support.md \
  docs/source_en/BestPractices/NPU-support.md
```

3. NPU Qwen GRPO smoke coverage:

| Scenario | Model | Backend / rollout | Training mode | Result |
| --- | --- | --- | --- | --- |
| colocate | Qwen3.5 dense | Megatron + vLLM-Ascend | full | 1 step passed, covering the basic dense + colocate path |
| colocate | Qwen3.5 dense | Megatron + vLLM-Ascend | LoRA | 1 step passed, covering dense + LoRA weight synchronization |
| colocate | Qwen3.5-MoE | Megatron + vLLM-Ascend | full | 1 step passed, covering Megatron MoE processed-layout synchronization |
| colocate | Qwen3.5-MoE | Megatron + vLLM-Ascend | LoRA | 1 step passed, covering the Megatron MoE + LoRA path |
| colocate | Qwen3.5-MoE | FSDP2 + vLLM-Ascend | full | 1 step passed, covering FSDP2 fused expert / pre-process layout synchronization |
| colocate | Qwen3-MoE | FSDP2 + vLLM-Ascend, vLLM TP=4 | LoRA | 5/5 steps passed, covering qwen3_moe fused expert runtime sync and processed-layout writeback |
| server | Qwen3.5-MoE | Megatron + standalone vLLM-Ascend server | full | 1 step passed, covering standalone server-mode HCCL broadcast, full-reload post-processing, and MoE inference |

Not covered:

- GPU regression was not validated locally; however, the NPU / vLLM-Ascend-specific logic is isolated behind runtime branches.
- PEFT fused expert `target_parameters + lora_dropout > 0` is still intentionally unsupported; the current approach is to document the limitation rather than maintain a temporary workaround in SWIFT.

### CODE REVIEW

**Comment by Jintao-Huang** [comment]
cc @hjh0119

**Review by hjh0119 — APPROVED** [review]

File: swift/model/npu_patch/vllm_ascend_moe.py [hunk 1]
```diff
@@ -126,7 +132,48 @@ def patch_vllm_ascend_moe_runtime() -> None:
...
 
 
-def expand_fused_moe_expert_names_for_vllm_ascend(name: str):
+def _is_qwen_moe_model(model) -> bool:
+    return getattr(getattr(model, 'config', None), 'model_type', None) in _QWEN_MOE_MODEL_TYPES
+
+
+def should_keep_fused_moe_expert_for_vllm_ascend(model) -> bool:
+    """Return whether fused expert names should be kept for vLLM-Ascend sync."""
+    return False
```
Comment (hjh0119) [inline swift/model/npu_patch/vllm_ascend_moe.py]: What is the purpose of this method?

File: swift/rlhf_trainers/vllm_client.py [hunk 2]
```diff
@@ -37,6 +38,15 @@
 logger = logging.getLogger(__name__)
 
 
+def _broadcast_tensor(communicator, tensor: torch.Tensor, src: int) -> None:
+    if is_torch_npu_available():
+        device_module = get_torch_device()
+        with device_module.device(communicator.device):
+            communicator.broadcast(tensor, src=src, stream=device_module.current_stream())
+    else:
+        communicator.broadcast(tensor, src=src, stream=getattr(get_torch_device(), 'current_stream', lambda: None)())
```
Comment (hjh0119) [inline swift/rlhf_trainers/vllm_client.py]: Consider reusing the methods from swift/pipelines/infer/rollout.py or extracting them into utils

File: swift/infer_engine/vllm_engine.py [hunk 3]
```diff
@@ -88,6 +88,33 @@ def _wait_for_zmq_addrs(self, zmq_addr_pipe):
...
+    vLLM 0.18.x Qwen3.5 configs may pass ``ignore_keys_at_rope_validation``
+    as a list, while Transformers 5.x treats it as a set and performs a set
+    union during RoPE validation. vLLM release tags from 0.19.0 onward changed
+    the Qwen3.5 configs to set literals, but 0.18-based vLLM/vLLM-Ascend stacks
+    still need this compatibility layer. See vLLM PR:
+    https://github.com/vllm-project/vllm/pull/37338
+    """
+    from transformers import PretrainedConfig
+
+    origin_convert = PretrainedConfig.convert_rope_params_to_dict
```
Comment (hjh0119) [inline swift/infer_engine/vllm_engine.py]: Agree with Gemini. Is this attribute safe to get for lower versions of Transformers

**Review by hjh0119 — APPROVED** [review]

**Comment by hjh0119** [comment]
Sorry for the late review. I've left a few minor comments

File: swift/model/npu_patch/vllm_ascend_moe.py (hunk 1 above)
Comment (addsubmuldiv) [reply swift/model/npu_patch/vllm_ascend_moe.py]: Thanks for pointing this out. This helper was a leftover from an earlier branch where we tried both "keep fused expert names" and "expand fused expert names" paths. After the final fix, vLLM-Ascend always needs the expanded checkpoint-style names for this sync path, so the helper had degenerated into a constant `False`, which was unnecessary and confusing.

I removed the helper and now call the fused MoE expansion path directly. This keeps the intent explicit and avoids carrying a dead switch in the code.

File: swift/rlhf_trainers/vllm_client.py (hunk 2 above)
Comment (addsubmuldiv) [reply swift/rlhf_trainers/vllm_client.py]: Done. I extracted the shared tensor broadcast logic into `broadcast_tensor_for_vllm_weight_sync()` in `swift/rlhf_trainers/utils.py`, and both `swift/rlhf_trainers/vllm_client.py` and `swift/pipelines/infer/rollout.py` now reuse it.

File: swift/infer_engine/vllm_engine.py (hunk 3 above)
Comment (addsubmuldiv) [reply swift/infer_engine/vllm_engine.py]: Fixed. The patch now uses `getattr(PretrainedConfig, "convert_rope_params_to_dict", None)` and skips the monkey patch if the method does not exist.

**Comment by hjh0119** [comment]
Thanks! Is this pr ready to merge?

**Comment by addsubmuldiv** [comment]
> Thanks! Is this pr ready to merge?

Yes, I think it’s ready. Please feel free to merge it if there are no further concerns.
